File: apps/go-example/wrapped_steps_story_test.go
package example

import (
	"testing"

	es "github.com/anthropics/executable-stories/packages/executable-stories-go"
)

func TestWrappedStepsAddition(t *testing.T) {
	s := es.Init(t, "Calculator adds two numbers using Fn and Expect")

	var a, b int
	s.Fn("Given", "two numbers 5 and 3", func() {
		a, b = 5, 3
	})

	var result int
	s.Fn("When", "the numbers are added", func() {
		result = Add(a, b)
	})

	s.Expect("the result is 8", func() {
		if result != 8 {
			t.Fatalf("expected 8, got %d", result)
		}
	})
}

func TestWrappedStepsSubtraction(t *testing.T) {
	s := es.Init(t, "Calculator subtracts using Fn with timing")

	var a, b int
	s.Fn("Given", "two numbers 10 and 4", func() {
		a, b = 10, 4
	})

	var result int
	s.Fn("When", "the second is subtracted from the first", func() {
		result = Subtract(a, b)
	})

	s.Expect("the result is 6", func() {
		if result != 6 {
			t.Fatalf("expected 6, got %d", result)
		}
	})
}

func TestWrappedStepsDivisionByZero(t *testing.T) {
	s := es.Init(t, "Calculator division by zero captured in Fn")

	s.Fn("Given", "a number 10 and zero", func() {})

	s.Expect("division by zero panics", func() {
		defer func() {
			if r := recover(); r == nil {
				t.Fatal("expected Divide(10, 0) to panic")
			}
		}()
		Divide(10, 0)
	})
}

func TestWrappedStepsMixed(t *testing.T) {
	s := es.Init(t, "Mixed markers and wrapped steps")

	s.Given("the calculator is ready")

	var result int
	s.Fn("When", "we multiply 7 by 6", func() {
		result = Multiply(7, 6)
	})

	s.Expect("the result is 42", func() {
		if result != 42 {
			t.Fatalf("expected 42, got %d", result)
		}
	})

	s.And("the result is a positive number")
	if result <= 0 {
		t.Fatal("expected positive number")
	}
}

File: apps/vitest-example/src/wrapped-steps.story.test.ts
import { story } from 'executable-stories-vitest';
import { describe, expect, it } from 'vitest';
import { add, divide, multiply, subtract } from './calculator.js';

describe('Wrapped Steps (fn/expect)', () => {
  it('Calculator adds two numbers using fn and expect', ({ task }) => {
    story.init(task);

    const a = story.fn('Given', 'number a is 5', () => 5);
    const b = story.fn('Given', 'number b is 3', () => 3);

    const result = story.fn('When', 'the numbers are added', () => add(a, b));

    story.expect('the result is 8', () => {
      expect(result).toBe(8);
    });
  });

  it('Calculator subtracts using fn with timing', ({ task }) => {
    story.init(task);

    const nums = story.fn('Given', 'two numbers 10 and 4', () => ({
      a: 10,
      b: 4,
    }));

    const result = story.fn('When', 'the second is subtracted from the first', () =>
      subtract(nums.a, nums.b),
    );

    story.expect('the result is 6', () => {
      expect(result).toBe(6);
    });
  });

  it('Calculator division by zero captured in fn', ({ task }) => {
    story.init(task);

    story.fn('Given', 'a number 10 and zero', () => {});

    story.expect('division by zero throws an error', () => {
      expect(() => divide(10, 0)).toThrow('Cannot divide by zero');
    });
  });

  it('Mixed markers and wrapped steps', ({ task }) => {
    story.init(task);

    story.given('the calculator is ready');

    const result = story.fn('When', 'we multiply 7 by 6', () => multiply(7, 6));

    story.expect('the result is 42', () => {
      expect(result).toBe(42);
    });

    story.and('the result is a positive number');
    expect(result).toBeGreaterThan(0);
  });

  it('Async fn wraps async operations with timing', async ({ task }) => {
    story.init(task);

    const data = await story.fn('Given', 'data fetched asynchronously', async () => {
      await new Promise((resolve) => setTimeout(resolve, 10));
      return { a: 5, b: 3 };
    });

    const result = await story.fn('When', 'async addition is performed', async () => {
      await new Promise((resolve) => setTimeout(resolve, 5));
      return add(data.a, data.b);
    });

    await story.expect('the async result is 8', async () => {
      expect(result).toBe(8);
    });
  });
});
